src/client/address_book/impl.cc

```diff
@@ -441,6 +452,31 @@ std::unique_ptr<const kovri::core::IdentHash> AddressBook::GetLoadedAddressIdent
   return nullptr;
 }
 
+void AddressBook::InsertAddress(
+    const std::string& host,
+    const kovri::core::IdentHash& address)
+{
+  try
+  {
+    // Ensure address book only inserts unique entries
+    if (!m_Addresses.empty())
+      {
+        const auto it = m_Addresses.find(host);
+        if (it != m_Addresses.end())
+          throw std::runtime_error("AddressBook: host already loaded");
```

Comment:
>const auto it = m_Addresses.find(host);
>if (it != m_Addresses.end())
>  throw std::runtime_error("AddressBook: host already loaded");

So, not only is this incredibly expensive, but it's not necessary because of `m_Addresses[host] = address`

`std::map::operator[] returns a reference to the value that is mapped to a key equivalent to key, performing an insertion if such key does not already exist.`

A much better solution is to tackle file writing / address saving *after* the map is properly updated.

Reply:
I didn't realize `std::map` worked that way, pretty cool feature. Removed the quoted block, but kept the check for loaded addresses. Eventually, likely after multiple subscriptions PR, there should be a reverse `address => hostname` lookup, making the current naive lookup unnecessary.

> A much better solution is to tackle file writing / address saving after the map is properly updated.

I will keep this in mind while working on the multiple subscription PR. Do we want to follow Java I2P naming for filenames, i.e. userhosts.txt, privatehosts.txt?

Reply:
>Do we want to follow Java I2P naming for filenames, i.e. userhosts.txt, privatehosts.txt?

There's no need, and `userhosts.txt` is also confusing.